lib/Authen/SASL/Perl/OAUTHBEARER.pm

```diff
@@ -0,0 +1,123 @@
+# Copyright (c) 2025 Aditya Garg <gargaditya08@live.com>
+# Copyright (c) 2025 Julian Swagemakers <julian@swagemakers.org>
+# All rights reserved.
+# This program is free software; you can redistribute it and/or
+# modify it under the same terms as Perl itself.
+
+package Authen::SASL::Perl::OAUTHBEARER;
+
+use strict;
+use vars qw(@ISA);
+
+@ISA     = qw(Authen::SASL::Perl);
+
+my %secflags = (
+    noanonymous => 1,
+);
+
+sub _order { 1 }
+
+sub _secflags {
+    shift;
+    scalar grep { $secflags{$_} } @_;
+}
+
+sub mechanism { 'OAUTHBEARER' }
+
+sub client_start {
+    my $self = shift;
+    $self->{stage} = 0;
+
+    # This will generate the oauthbearer string used for authentication.
+    #
+    # "n,a=" {User} ",^Ahost=" {Host} "^Aport=" {Port} "^Aauth=Bearer " {Access Token} "^A^A
+    #
+    # The first part `n,a=" {User} ",` is the gs2 header described in RFC5801.
+    # * gs2-cb-flag `n` -> client does not support CB
+    # * gs2-authzid `a=" {User} "`
+    #
+    # The second part are key value pairs containing host, port and auth as
+    # described in RFC7628.
+    #
+    # https://datatracker.ietf.org/doc/html/rfc5801
+    # https://datatracker.ietf.org/doc/html/rfc7628
+    my $username = $self->_call('user');
+    my $token    = $self->_call('pass'); # OAuth 2.0 access token
+    my $auth_string = "n,a=$username,\001host=$self->{host}\001port=$self->{port}\001auth=Bearer $token\001\001";
+
+    return $auth_string;
+}
+
+sub client_step {
+    my ($self, $challenge) = @_;
+
+    my $stage = ++$self->{stage};
+
+    if ($stage == 1) {
+        # Send dummy request on authentication failure according to rfc7628.
+        # https://datatracker.ietf.org/doc/html/rfc7628#section-3.2.3
+        return "\001";
+    } else {
+        return $self->set_error("Invalid sequence");
+    }
+}
+
+1;
+
+__END__
+
+=head1 NAME
+
+Authen::SASL::Perl::OAUTHBEARER - OAUTHBEARER Authentication class
+
+=head1 SYNOPSIS
+
+  use Authen::SASL qw(Perl);
+
+  $sasl = Authen::SASL->new(
+    mechanism => 'OAUTHBEARER',
+    callback  => {
+      user => $user,
+      pass => $access_token
+    },
+  );
```

Comment:
Since the OAUTHBEARER mechanism also depends on the "host" and "port" variables, it would be good to include a comment in the code above that the user of the SASL object (the caller/connection owner) is supposed to set these properties on the connection, or if the caller should do that, include that in the code above.

Reply:
I'll have to add an additional port option

Reply:
I guess removing the use of the port below resolves this comment too.